consensus/core/src/base_committer.rs

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    block::{Block, BlockAPI, BlockRef},
+    block_store::BlockStore,
+    constants::{DEFAULT_WAVE_LENGTH, MINIMUM_WAVE_LENGTH},
+    leader_schedule::LeaderSchedule,
+    stake_aggregator::{QuorumThreshold, StakeAggregator},
+    types::{AuthorityRound, LeaderStatus, Round, WaveNumber},
+};
+
+use std::{collections::HashMap, fmt::Display, sync::Arc};
+
+use consensus_config::{AuthorityIndex, Committee, Stake};
+
+pub struct BaseCommitterOptions {
+    /// TODO: Re-evaluate if we want this to be configurable after running experiments.
+    /// The length of a wave (minimum 3)
+    pub wave_length: u32,
+    /// The offset used in the leader-election protocol. This is used by the
+    /// multi-committer to ensure that each [`BaseCommitter`] instance elects
+    /// a different leader.
+    pub leader_offset: u32,
+    /// The offset of the first wave. This is used by the pipelined committer to
+    /// ensure that each[`BaseCommitter`] instances operates on a different
+    /// view of the dag.
+    pub round_offset: u32,
+}
+
+impl Default for BaseCommitterOptions {
+    fn default() -> Self {
+        Self {
+            wave_length: DEFAULT_WAVE_LENGTH,
+            leader_offset: 0,
+            round_offset: 0,
+        }
+    }
+}
+
+/// The [`BaseCommitter`] contains the bare bone commit logic. Once instantiated,
+/// the method `try_direct_decide` and `try_indirect_decide` can be called at any
+/// time and any number of times (it is idempotent) to determine whether a leader
+/// can be committed or skipped.
+#[allow(unused)]
+pub struct BaseCommitter {
+    /// The committee information
+    committee: Arc<Committee>,
+    /// The consensus leader schedule to be used to resolve the leader for a
+    /// given round.
+    leader_schedule: LeaderSchedule,
+    /// Keep all block data
+    block_store: BlockStore,
+    /// The options used by this committer
+    options: BaseCommitterOptions,
+}
+
+impl BaseCommitter {
+    pub fn new(
+        committee: Arc<Committee>,
+        leader_schedule: LeaderSchedule,
+        block_store: BlockStore,
+    ) -> Self {
+        Self {
+            committee,
+            leader_schedule,
+            block_store,
+            options: BaseCommitterOptions::default(),
+        }
+    }
+
+    pub fn with_options(mut self, options: BaseCommitterOptions) -> Self {
+        assert!(options.wave_length >= MINIMUM_WAVE_LENGTH);
+        self.options = options;
+        self
+    }
```

Comment:
Don't have an issue with it, but maybe for now we can make the `options` be passed as part of the `new` method to minimise any possibility for misconfiguration as it is a critical piece of code and introduce the `with_options` later if needed.